### DISCUSSION TITLE
HTTP Early Hints PoC

### DISCUSSION THREAD

**Discussion by arut** [nginx/nginx#332] [Announcements]
In PR #326 we have a proof-of-concept implementation of [HTTP Early Hints](https://datatracker.ietf.org/doc/html/rfc8297). We've been receiving multiple requests for this feature and now we are proud to share this PoC with the community. For anyone who's interested, please try it and report any feedback. This includes syntax, implementation issues, performance issues etc.

**Comment by ZeusF1** [comment]
Any estimates on merge?

**Comment by totoro-andriy** [comment]
Hi Roman! Thanks a lot for the great patch 🔥

We've been using your implementation in production on a high-traffic project — everything works smoothly, with no performance issues observed.

For some specific needs, we built an extension based on your fork: [README.early-hints.md](https://github.com/totoro-andriy/nginx/blob/master/README.early-hints.md)

Solid work, it was a pleasure to integrate. Thanks again!

**Comment by L1H0n9Jun** [comment]
Excellent feature! 

Does it work on the current HEAD (f3542500b6d74e3e88fc84b88144afe67882d1fa)? 
Applied the patch on top of the latest commit. The `Link` header appears, but no 103 response received.

Not sure if there is a configuration error:

```conf
server {
    server_name  example.com;
    listen       443      quic;
    listen       [::]:443 quic;
    listen       443      ssl;
    listen       [::]:443 ssl;

    include      conf.d/base/ssl.conf;
    include      conf.d/base/header.conf;

    root         /var/www/;
    index        index.html;

    location = /index.html {
        include     conf.d/base/header.conf;

        add_header  Link "</css/main.css>; rel=preload; as=style" early;

        early_hints $http2 $http3;
    }
}
```

conf.d/base/header.conf:
```conf
add_header  Alt-Svc 'h3=":443"; ma=86400';
add_header  Cache-Control "public, max-age=604800";
add_header  Strict-Transport-Security "max-age=31536000; includeSubdomains; preload";
add_header  X-Frame-Options "SAMEORIGIN" always;
add_header  X-XSS-Protection "1; mode=block" always;
add_header  X-Content-Type-Options "nosniff" always;
```

Let me know if more configuration details are needed.

Thanks for your excellent work!

**Reply by totoro-andriy** [reply]
the Early Hints were only successfully emitted when I had a proxy_pass in place (like to a backend service), not when Nginx served static files directly from disk.

I suspect the same may be happening in your case: the Link header is set (good), but since there’s no upstream to delay the final response, Nginx just proceeds straight to the full 200 without bothering to emit 103.

`location / {

        add_header Link "</css/main.css>; rel=preload; as=style" early;
		early_hints $http2 $http3;

        proxy_pass http://127.0.0.1:8192;
        proxy_http_version 1.1;
        proxy_set_header Scheme $scheme;
        proxy_set_header Https on;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header SERVER_PORT $server_port;
        proxy_set_header REMOTE_ADDR $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade;
        proxy_read_timeout     120;
        proxy_connect_timeout  120;
}`

**Reply by L1H0n9Jun** [reply]
> the Early Hints were only successfully emitted when I had a proxy_pass in place (like to a backend service), not when Nginx served static files directly from disk.
> 
> I suspect the same may be happening in your case: the Link header is set (good), but since there’s no upstream to delay the final response, Nginx just proceeds straight to the full 200 without bothering to emit 103.
> 
> `location / {
> 
> ```
>     add_header Link "</css/main.css>; rel=preload; as=style" early;
> 	early_hints $http2 $http3;
> 
>     proxy_pass http://127.0.0.1:8192;
>     proxy_http_version 1.1;
>     proxy_set_header Scheme $scheme;
>     proxy_set_header Https on;
>     proxy_set_header Host $host;
>     proxy_set_header X-Forwarded-Proto $scheme;
>     proxy_set_header SERVER_PORT $server_port;
>     proxy_set_header REMOTE_ADDR $remote_addr;
>     proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
>     proxy_set_header Upgrade $http_upgrade;
>     proxy_set_header Connection $connection_upgrade;
>     proxy_read_timeout     120;
>     proxy_connect_timeout  120;
> ```
> 
> }`

Absolutely, it works when Nginx is used as a reverse proxy.

Not sure why 103 doesn't take effect when Nginx serves static sites. For some complex static sites, 103 should have an optimization effect on the loading of blocking resources? Are there any considerations?

**Reply by pluknet** [reply]
> Not sure why 103 doesn't take effect when Nginx serves static sites.

For static files, you can use plain `add_header Link ...;` for such an optimization, e.g. in a (rather unrealistic) scenario when a static html file appears to be very large and it contains some links closer to its end.
There's no point here in signalling using early hints, because response headers for static files are sent immediately.

**Comment by anthonyryan1** [comment]
I was beyond excited to see early hints in nginx!

I've been waiting years for this, formerly an excited user of HTTP/2 Server Push before it got misused by developers and the browsers axed it.

I must admit I was a bit devastated to find out that the ability to have nginx generate early hints was removed from the scope. I can respect working on the feature incrementally, but from my perspective we are still missing the most valuable and interesting part. 

If there's an open issue to track development on the ability for nginx to add early hints instead of just forwarding them as a proxy. Please let me know where I would very much like to subscribe to it.

-----

My use case is:

A PHP FPM FastCGI backend. As far as I can tell, nearly all implementations of the FastCGI protocol have assumptions baked in that prevent HTTP 103 at that level. They assume one HTTP staus per request. (I'm not even certain the protocol allows something like 103 Early Hints to exist).

I would like to have nginx send early_hints for CSS, images above the fold, and (if any) render blocking JavaScript to the browser, so it can start fetching those resources. Basically everything that's needed between receiving HTML and First Contentful Paint. So that the browser can fetch those resources in parallel while PHP does any server side work to generate the page.

By doing both of these in parallel, a web page can load an entire RTT faster, for having the critical subresources ready before the HTML arrives.

Right now if there's 100 ms of RTT, and 100ms of server side computation to render a page. A browser will take 300 ms to show the page the user. 100 ms for the initial request, 100 ms for server side computation, and 100 ms to fetch css, images, and blocking js.

If nginx could send early hints, this same scenario could be reduced to 200 ms. 100 ms for the initial request, the browser receives the early hints from nginx, when it would otherwise be idle, and at the end of the server side computation instantly shows the page.

This is a big difference for people accessing web sites hosted physically far away across the world. Allowing us reduce their page load times from (2x RTT + server processing time) to (1x RTT + server processing time).

For people on the far side of the world (from the web server), with bad cellular or satellite internet. This is a *huge* difference in how fast a web page loads.

-----

From a semantic correctness standpoint, I think I'd prefer to specify Early Hints at the PHP layer, rather than the Nginx layer. I'm just not confident it can be done while conforming to the FastCGI spec.

**Reply by arut** [reply]
@anthonyryan1 thanks for your feedback.

Regarding FastCGI + early hints, [there's a chance](https://github.com/nginx/nginx/pull/326#issuecomment-2847467043).

**Reply by L1H0n9Jun** [reply]
@anthonyryan1 Just add preload Link headers (without 103 early hints) at nginx layer:

```
add_header    Link "</path-to-script.js>; rel=preload; as=script; crossorigin";
add_header    Link "</path-to-stylesheet.css>; rel=preload; as=style";
```

and see if it works as you expected (Resources are directly requested with priority).

**Reply by anthonyryan1** [reply]
I put together a quick proof of concept showing how that helps and why it's not quite as good as Early Hints.

This page has 1 CSS, 1 Image, 1 Image loaded from the CSS, 200 ms of latency and an artificial 3000 ms sleep in the HTML generation to emulate server computation time.

The baseline doing nothing:
![Early Hints Baseline](https://github.com/user-attachments/assets/18647d93-90f7-449b-af1c-263e7c7452cd)

Your suggestion of using Link preload allows us to cut the RTT of sub-subresources. Like an image in CSS, but doesn't do anything to load the images before the HTML returns, because the Link header doesn't arrive until the server is done all processing.
![Early Hints Link Preload](https://github.com/user-attachments/assets/2fb7c560-2765-4fd1-b378-aaa4696c6a1f)

103 Early Hints would put the CSS and both images at the start of the server computation period, rather than starting them immediately after it returns, still chopping the RTT I was talking about.

I hope these waterfall charts help clarify the benefit.

**Reply by Ghostavio** [reply]
[anthonyryan1](https://github.com/anthonyryan1) wait so this is not currently supported by the 1.29.0 release yet?

**Reply by anthonyryan1** [reply]
@Ghostavio My understanding and experience so far, is that nginx in 1.29.0 now supports acting as a reverse proxy for an upstream which already supports early hints. But cannot add early hints itself.

**Reply by Ghostavio** [reply]
@anthonyryan1 then their [blog post](https://blog.nginx.org/blog/nginx-introduces-support-103-early-hints) was very misleading to say the least.

**Comment by Ghostavio** [comment]
I don't know if I'm doing something wrong, I've added the following to the top of my `/etc/nginx/sites-enabled/site` file:
```
map $http_sec_fetch_mode $early_hints {
    navigate $http2$http3;
}
```

And `early_hints $early_hints;` inside `location` within `server`, just before my `proxy_pass`, as instruced by the blog post. Then I proceeded to add the `Link` header with the content I want to have loaded by the Early Hints while my html is loading, following the proper format `</path/to/file> rel=preload; as=script; crossorigin, </path/to/another-file> rel=preload; as=style; crossorigin`. This is being served by node.js for an SSR application through nginx 1.29.0.

This has done absolutely nothing, I don't get the Early Hints showing on Chrome Dev Tools, testing my site [here](https://code103.hotmann.de/) says **No "Early Hints" has been found!**. What am I missing? The blog post is saying a lot about this amazing feature but what else do I need to do to make it work?

This is my entire `location` object:
```
location / {
        early_hints $early_hints;
        proxy_pass http://localhost:3060;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
}
```

**Reply by binaryfire** [reply]
Which blog post are you referring to?

**Reply by Ghostavio** [reply]
https://blog.nginx.org/blog/nginx-introduces-support-103-early-hints

**Reply by Ghostavio** [reply]
Can anyone help with this?

**Reply by MarcoMarcoaldi** [reply]
Same trouble. No guidelines how to add .css early hints.

**Comment by Uzume** [comment]
I am linking in this issue from Vite: vikejs/vike#1452

**Reply by johaannaburkhardt1995-ops** [reply]
Liste meiner Erbe
